hydra-main/src/main/java/com/addthis/hydra/job/alias/AliasManagerImpl.java

```diff
@@ -51,14 +99,158 @@ public AliasManagerImpl(SpawnDataStore spawnDataStore) {
+            log.warn("failed to put alias: {}", alias, e);
+            throw Throwables.propagate(e);
+        } finally {
+            mapLock.unlock();
+        }
+    }
+
+    /**
+     * Get all jobIds for a given alias
+     *
+     * @param alias The alias to check
+     * @return A list of jobIds, possible null
+     */
+    public List<String> getJobs(String alias) {
+        refreshAlias(alias);
+        mapLock.lock();
+        try {
+            return alias2jobs.get(alias);
+        } finally {
+            mapLock.unlock();
+        }
+    }
+
+    /**
+     * Refresh an alias based on datastore
+     *
+     * @param alias The alias to refresh
+     */
+    private void refreshAlias(String alias) {
+        try {
+            updateAlias(alias, this.spawnDataStore.getChild(ALIAS_PATH, alias));
+        } catch (ExecutionException e) {
+            log.warn("Failed to refresh alias: {}", alias, e);
+        } catch (Exception e) {
+            log.error("",e);
+        }
+    }
+
+    /**
+     * Load the jobIds for a particular alias from the SpawnDataStore
+     *
+     * @param alias The alias key to check
+     * @return String The data that was updated (so the cache can be updated)
+     */
+    @Nullable private String updateAlias(String alias, @Nullable String data) {
+        if (Strings.isNullOrEmpty(alias)) {
+            return data;
+        }
+        if (Strings.isNullOrEmpty(data)) {
+            deleteAlias(alias);
+            return data;
+        }
+        List<String> jobs = decodeAliases(data);
+        if (jobs.isEmpty()) {
+            log.warn("no jobs for alias {}, ignoring {}", alias, alias);
+            return data;
+        }
+        mapLock.lock();
+        try {
+            alias2jobs.put(alias, jobs);
+            job2alias.put(jobs.get(0), alias);
+        } finally {
+            mapLock.unlock();
+        }
+        return data;
+    }
+
+    @VisibleForTesting
+    protected List<String> decodeAliases(@Nonnull Object data) {
+        try {
+            return mapper.readValue(data.toString(), new TypeReference<List<String>>() {});
+        } catch (IOException e) {
+            log.warn("Failed to decode data", e);
+            return new ArrayList<>(0);
+        }
+    }
+
+    /**
+     * Delete the data for a given alias
+     *
+     * @param alias The alias to check
+     */
     public void deleteAlias(String alias) {
-        aliasBiMap.deleteAlias(alias);
+        mapLock.lock();
+        try {
+            List<String> jobs = alias2jobs.get(alias);
+            alias2jobs.remove(alias);
+            if ((jobs != null) && !jobs.isEmpty()) {
+                for (String job : jobs) {
+                    String aliasVal = job2alias.get(job);
+                    if (Objects.equals(aliasVal, alias)) {
+                        job2alias.remove(job);
+                    }
+                }
+            }
+        } finally {
+            mapLock.unlock();
+        }
```

Comment:
Shouldn't you copy the maps first? What happens if the `deleteChild` call below fails? You now have a bad state.